io/src/main/scala/sbt/io/IO.scala

```diff
@@ -551,16 +552,16 @@ object IO {
   }
 
   /** Deletes `file`, recursively if it is a directory. */
-  def delete(file: File): Unit = retry {
+  def delete(file: File): Unit = Retry {
     try {
-      FileTreeView.DEFAULT.list(Glob(file, AllPassFilter, 0)).foreach {
-        case dir if dir.isDirectory => delete(dir.toPath.toFile)
-        case f                      => f.toPath.toFile.delete()
+      FileTreeView.default.list(Glob(file, AnyPath)).foreach {
+        case (dir, attrs) if attrs.isDirectory => delete(dir.toFile)
+        case (f, _)                            => Files.deleteIfExists(f)
```

Comment:
https://jenkins.scala-sbt.org:8592/job/sbt-validator/1569/consoleFull

```
[sbt212-librarymanagement] [info] - should reuse the caches when offline and cached resolution are enabled *** FAILED ***
[sbt212-librarymanagement] [info]   java.nio.file.DirectoryNotEmptyException: ./target/OfflineModeSpec/ivyhome/cache
[sbt212-librarymanagement] [info]   at sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:242)
[sbt212-librarymanagement] [info]   at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
[sbt212-librarymanagement] [info]   at java.nio.file.Files.deleteIfExists(Files.java:1165)
[sbt212-librarymanagement] [info]   at sbt.io.IO$.$anonfun$delete$2(IO.scala:564)
[sbt212-librarymanagement] [info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[sbt212-librarymanagement] [info]   at sbt.internal.io.Retry$.liftedTree2$1(Retry.scala:38)
[sbt212-librarymanagement] [info]   at sbt.internal.io.Retry$.impl$1(Retry.scala:38)
[sbt212-librarymanagement] [info]   at sbt.internal.io.Retry$.apply(Retry.scala:50)
[sbt212-librarymanagement] [info]   at sbt.internal.io.Retry$.apply(Retry.scala:24)
[sbt212-librarymanagement] [info]   at sbt.io.IO$.delete(IO.scala:555)
```

The error is `java.nio.file.DirectoryNotEmptyException: ./target/OfflineModeSpec/ivyhome/cache`. Does this point to the issue of caching?

Reply:
No, the issue is that FileTreeView.list(glob: Glob) didn't handle relative globs correctly. I both fixed the issue with relative globs and also changed this line from `FileTreeView.default.list(Glob(file, AnyPath)).foreach {` to `FileTreeView.default.list(file.toPath).foreach {` which should be semantically equivalent but has a simpler implementation.